sycl/source/detail/syclbin.cpp

```diff
@@ -0,0 +1,393 @@
+  uint64_t NumStrings;
+  uint64_t ImageOffset;
+  uint64_t ImageSize;
+};
+
+class BlockReader {
+protected:
+  BlockReader(const char *Data, size_t Size) : Data{Data}, Size{Size} {}
+
+  void ReadSizeCheck(size_t ByteOffset, size_t ReadSize) {
+    if (ByteOffset + ReadSize > Size)
+      throw sycl::exception(make_error_code(errc::invalid),
+                            "Unexpected file contents size.");
+  }
+
+  const char *Data = nullptr;
+  size_t Size = 0;
+};
+
+class HeaderBlockReader : public BlockReader {
+public:
+  HeaderBlockReader(const char *Data, size_t Size) : BlockReader(Data, Size) {}
+
+  template <typename HeaderT> const HeaderT *GetHeaderPtr(size_t ByteOffset) {
+    ReadSizeCheck(ByteOffset, sizeof(HeaderT));
+    return reinterpret_cast<const HeaderT *>(Data + ByteOffset);
+  }
+};
+
+class SYCLBINByteTableBlockReader : public BlockReader {
+public:
+  SYCLBINByteTableBlockReader(const char *Data, size_t Size)
+      : BlockReader(Data, Size) {}
+
+  std::string_view GetBinaryBlob(size_t ByteOffset, uint64_t BlobSize) {
+    ReadSizeCheck(ByteOffset, BlobSize);
+    return {Data + ByteOffset, BlobSize};
+  }
+
+  std::unique_ptr<PropertySetRegistry> GetMetadata(size_t ByteOffset,
+                                                   uint64_t MetadataSize) {
+    return PropertySetRegistry::read(GetBinaryBlob(ByteOffset, MetadataSize));
+  }
+};
+
+std::pair<const char *, size_t> getImageInOffloadBinary(const char *Data,
+                                                        size_t Size) {
+  if (sizeof(OffloadBinaryHeaderType) > Size)
+    throw sycl::exception(make_error_code(errc::invalid),
+                          "Invalid Offload Binary size.");
+
+  // Read the header.
+  const OffloadBinaryHeaderType *Header =
+      reinterpret_cast<const OffloadBinaryHeaderType *>(Data);
+  if (memcmp(Header->Magic, OffloadBinaryMagic, 4) != 0)
+    throw sycl::exception(make_error_code(errc::invalid),
+                          "Incorrect Offload Binary magic number.");
+
+  if (Header->Version != 1)
+    throw sycl::exception(make_error_code(errc::invalid),
+                          "Unsupported Offload Binary version number.");
+
+  if (Header->Version != 1)
+    throw sycl::exception(make_error_code(errc::invalid),
+                          "Unsupported Offload Binary version number.");
+
+  if (Header->EntrySize != sizeof(OffloadBinaryEntryType))
+    throw sycl::exception(make_error_code(errc::invalid),
+                          "Unexpected number of offload entries.");
+
+  if (Header->EntryOffset + sizeof(OffloadBinaryEntryType) > Size)
+    throw sycl::exception(make_error_code(errc::invalid),
+                          "Invalid entry offset.");
+
+  // Read the table entry.
+  const OffloadBinaryEntryType *Entry =
+      reinterpret_cast<const OffloadBinaryEntryType *>(Data +
+                                                       Header->EntryOffset);
+
+  if (Entry->ImageKind != /*IMG_SYCLBIN*/ 6)
```

Comment:
Should we move this constant somewhere? Where did `6` come from?

Reply:
It's sadly a little bit of a hack, stemming from https://github.com/intel/llvm/blob/sycl/llvm/include/llvm/Object/OffloadBinary.h#L50. Eventually this file will disappear in favor of the LLVMObject version of SYCLBIN parsing, but for now we can't access that enum. Do you think it would make sense to make a `constexpr uint16_t IMG_SYCLBIN = 6;` and use that? I'm not sure how much it adds as it's still just a magic value, but maybe it's easier to make the intention clear. Alternatively, I can add a comment.

Reply:
Do you think this value will be used somewhere else in this temporary file? If not, I guess this is fine to keep as is, otherwise constexpr variable is good to be added, but overall this is nit